File: 15_CoreDataDemo/CoreDataTool_oc/CoreDataTool_oc/CoreDataTool/CoreDataTool.m
//
//  CoreDataTool.m
//  CoreDataTool_oc
//
//  Created by sunke on 2020/9/20.
//  Copyright © 2020 KentSun. All rights reserved.
//

#import "CoreDataTool.h"
#import "AppDelegate.h"
#import <CoreData/CoreData.h>

// 如果项目创建时未勾选CoreData, 将此值改为 0
#define CoreDataToolAuto 1
//static CoreDataTool * __coreData = nil;
// .xcdatamodeld 文件名称
static NSString *const CoreDataToolModelName = @"CoreDataTool_oc";

@implementation CoreDataTool

+ (NSManagedObjectContext *)coreDataContext {
    
#ifdef CoreDataToolAuto
    AppDelegate *delegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    return delegate.persistentContainer.viewContext;
#else
    // 如果是新建的CoreDataToolModel.xcdatamodeld, 则需要手动加载
    NSPersistentContainer *container = [[NSPersistentContainer alloc]initWithName:CoreDataToolModelName];
    [container loadPersistentStoresWithCompletionHandler:^(NSPersistentStoreDescription * storeDescription, NSError * error) {
        if (error != nil) {
            NSLog(@"%@", error.userInfo);
        }
    }];
    
    return container.viewContext;
#endif
}

+ (void)saveContext {
    
#ifdef CoreDataToolAuto
    AppDelegate *delegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    [delegate saveContext];
#else
    NSManagedObjectContext *context = [self coreDataContext];
    NSError *error = nil;
    if ([context hasChanges] && ![context save:&error]) {
        // Replace this implementation with code to handle the error appropriately.
        // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
        NSLog(@"Unresolved error %@, %@", error, error.userInfo);
        abort();
    }
#endif
}

+ (void)insertEntity:(NSString *)name configData:(void(^)(NSManagedObject *obj))handle {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    NSManagedObject *obj = [NSEntityDescription insertNewObjectForEntityForName:name inManagedObjectContext:cxt];
    handle(obj);
    [self saveContext];
}

+ (void)fetchEntity:(NSString *)name predicate:(NSPredicate *)pred properties:(NSArray *)pro sortKey:(NSString *)key result:(void(^)(NSArray *objs))block {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    dispatch_queue_t queue = dispatch_queue_create("CoreDataTool_fetchQueue", NULL);
    dispatch_async(queue, ^{
        
        NSFetchRequest *req = [[NSFetchRequest alloc]init];
        NSEntityDescription *entity = [NSEntityDescription entityForName:name inManagedObjectContext:cxt];
        req.entity = entity;
        req.predicate = pred;
        if (pro) {
            req.propertiesToFetch = pro;
        }
        
        if (key) {
            NSSortDescriptor *sort = [NSSortDescriptor sortDescriptorWithKey:key ascending:YES];
            req.sortDescriptors = @[sort];
        }
        
        NSError *error;
        NSArray *rs = [cxt executeFetchRequest:req error:&error];
        if (error == nil) {
            if (block) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    block(rs);
                });
            }
        } else {
            NSLog(@"%@", error.userInfo);
        }
    });
}

+ (void)deleteEntity:(NSString *)name predicate:(NSPredicate *)pred result:(void(^)(NSArray *deletedObjs))block {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    
    NSFetchRequest *req = [[NSFetchRequest alloc]init];
    NSEntityDescription *entity = [NSEntityDescription entityForName:name inManagedObjectContext:cxt];
    req.entity = entity;
    req.predicate = pred;
    req.includesPropertyValues = NO;
    
    NSError *error;
    NSArray *deleteObjs = [cxt executeFetchRequest:req error:&error];
    if (error == nil) {
        for (NSManagedObject *obj in deleteObjs) {
            [cxt deleteObject:obj];
        }
        
        if (block) {
            block(deleteObjs);
        }
    } else {
        NSLog(@"%@", error.userInfo);
    }
    
    [self saveContext];
}

+ (void)updateEntity:(NSString *)name predicate:(NSPredicate *)pred configNewObj:(void(^)(NSArray *objs))objBlock result:(void(^)(NSError *error))rsBlock {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    NSFetchRequest *req = [[NSFetchRequest alloc]init];
    NSEntityDescription *entity = [NSEntityDescription entityForName:name inManagedObjectContext:cxt];
    req.entity = entity;
    req.predicate = pred;
    
    NSError *error;
    NSArray *rs = [cxt executeFetchRequest:req error:&error];
    objBlock(rs);
    [self saveContext];
    
    if (rsBlock) {
        rsBlock(error);
    }
}

+ (void)batchDeleteEntity:(NSString *)name predicate:(NSPredicate *)pred result:(void(^)(NSError *error))rsBlock {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    NSFetchRequest *req = [NSFetchRequest fetchRequestWithEntityName:name];
    req.predicate = pred;
    
    NSBatchDeleteRequest *batchDele = [[NSBatchDeleteRequest alloc]initWithFetchRequest:req];
    batchDele.resultType = NSBatchDeleteResultTypeObjectIDs;
    
    NSError *error;
    NSBatchDeleteResult *result = [cxt executeRequest:batchDele error:&error];
    
    if (result.result) {
        [NSManagedObjectContext mergeChangesFromRemoteContextSave:@{NSDeletedObjectsKey: result.result} intoContexts:@[cxt]];
    }
    
    if (rsBlock) {
        rsBlock(error);
    }
}

+ (void)batchUpdateEntity:(NSString *)name predicate:(NSPredicate *)pred propertiesToUpdate:(NSDictionary *)values result:(void(^)(NSError *error))rsBlock {
    
    NSManagedObjectContext *cxt = [self coreDataContext];
    NSBatchUpdateRequest *req = [NSBatchUpdateRequest batchUpdateRequestWithEntityName:name];
    req.predicate = pred;
    req.propertiesToUpdate = values;
    req.resultType = NSUpdatedObjectIDsResultType;
    
    NSError *error;
    NSBatchUpdateResult *rs = [cxt executeRequest:req error:&error];
    if (rs.result) {
        [NSManagedObjectContext mergeChangesFromRemoteContextSave:@{NSUpdatedObjectsKey: rs.result} intoContexts:@[cxt]];
    }
    
    if (rsBlock) {
        rsBlock(error);
    }
}

+ (NSPredicate *)predicateOfLike:(NSString *)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K LIKE %@", key, value];
}

+ (NSPredicate *)predicateOfMatch:(NSString *)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K MATCHES %@", key, value];
}

+ (NSPredicate *)predicateOfMatchc:(NSString *)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K MATCHES[c] %@", key, value];
}

+ (NSPredicate *)predicateOfContain:(NSString *)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K CONTAINS %@", key, value];
}

+ (NSPredicate *)predicateOfEqual:(NSInteger)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K == %d", key, value];
}

+ (NSPredicate *)predicateOfBigThan:(NSInteger)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K > %d", key, value];
}

+ (NSPredicate *)predicateOfSmallThan:(NSInteger)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K < %d", key, value];
}

+ (NSPredicate *)predicateOfNotSmallThan:(NSInteger)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K >= %d", key, value];
}

+ (NSPredicate *)predicateOfNotBigThan:(NSInteger)value forProperty:(NSString *)key {
    
    return [NSPredicate predicateWithFormat:@"%K LIKE %@", key, value];
}

@end

File: 15_CoreDataDemo/CoreDataTool_swift/CoreDataTool_swift/CoreDataTool/CoreDataTool.swift
//
//  CoreDataTool.swift
//  CoreDataTool_swift
//
//  Created by sunke on 2020/9/20.
//  Copyright © 2020 KentSun. All rights reserved.
//

import UIKit
import CoreData

fileprivate let CoreDataToolModelName = "CoreDataTool_swift"

class CoreDataTool {

    private static var context: NSManagedObjectContext = {
        // 如果是新建项目时启用了Core Data, 则可使用AppDelegate中的context
        // 即
        //        let delegate = UIApplication.shared.delegate as! AppDelegate
        //        let context = delegate.persistentContainer.viewContext
        //
        //        return context
        
        // 如果是新建的LDCoreDataModel.xcdatamodeld, 则需要手动加载
        let container = NSPersistentContainer(name: CoreDataToolModelName)
        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
            if let error = error as NSError? {
                
                fatalError("Unresolved error \(error), \(error.userInfo)")
            }
        })
        
        return container.viewContext
    }()
    
    private class func saveContext(_ resuleHandler: ((_ error: NSError?) -> Void)? = nil) {
        
        if context.hasChanges {
            do {
                try context.save()
                if let handler = resuleHandler {
                    handler(nil)
                }
            } catch {
                let nserror = error as NSError
                if let handler = resuleHandler {
                    handler(nserror)
                }
                fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
            }
        }
    }
    
    /// 增
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - handler: 配置待保存数据回调
    ///   - rs: 保存结果回调
    class func insert(entity name: String, configHandler handler: ((_ obj: NSManagedObject) -> Void), resulteHandler rs: ((_ error: NSError? ) -> Void)? = nil) {
        
        let obj = NSEntityDescription.insertNewObject(forEntityName: name, into: context)
        
        handler(obj)
        
        saveContext(rs)
    }
    
    /// 查
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - predicate: 查询条件-谓词
    ///   - propertiesToFetch: 查询的属性, 默认所有属性
    ///   - key: 查询结果排序的依据属性, 升序
    ///   - result: 查询结果回调
    class func fetch(entity name: String, predicate: NSPredicate? = nil, propertiesToFetch: Array<Any>? = nil, resultSortKey key: String? = nil, resultHandler rs: @escaping ((_ info: Array<Any>?) -> Void)) {
        
        let queue = DispatchQueue(label: "fetchQueue")
        queue.async {
            let fetchReq = NSFetchRequest<NSFetchRequestResult>()
            let entity = NSEntityDescription.entity(forEntityName: name, in: context)
            
            fetchReq.entity = entity
            fetchReq.predicate = predicate// 设置查询条件
            fetchReq.propertiesToFetch = propertiesToFetch// 设置查询属性, 默认查询全部
            if let key = key {
                let sort = NSSortDescriptor(key: key, ascending: true)
                fetchReq.sortDescriptors = [sort]
            }
            
            let fetchObjs = try? context.fetch(fetchReq)
            
            DispatchQueue.main.async {
                rs(fetchObjs)
            }
        }
    }
    
    /// 删
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - predicate: 删除条件
    ///   - result: 删除结果回调
    class func delete(entity name: String, predicate: NSPredicate? = nil, resultHandler rs: ((_ error: NSError?, _ deletedObjs: Array<Any>?) -> Void)? = nil) {
        
        let fetchReq = NSFetchRequest<NSFetchRequestResult>()
        let entity = NSEntityDescription.entity(forEntityName: name, in: context)
        
        fetchReq.entity = entity
        fetchReq.predicate = predicate // 设置查询条件
        fetchReq.includesPropertyValues = false
        
        do {
            let fetchedObjs = try context.fetch(fetchReq)
            
            for obj in fetchedObjs {
                context.delete(obj as! NSManagedObject)
            }
            
            try context.save()
            
            if let rs = rs {
                rs(nil, fetchedObjs)
            }
            
        } catch let error {
            if let rs = rs {
                rs(error as NSError, nil)
            }
        }
    }
    
    /// 改
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - predicate: 查询条件
    ///   - handler: 更新数据的回调
    ///   - result: 更新结果的回调
    class func update(withEntityName name: String, predicate: NSPredicate, configNewValues handler: ((_ objs: Array<Any>) -> Void), result: ((_ error: NSError?) -> Void)? = nil) {
        
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>()
        let entity = NSEntityDescription.entity(forEntityName: name, in: context)
        
        fetchRequest.entity = entity
        fetchRequest.predicate = predicate // 设置查询条件
        
        do {
            let fetchedObjs = try context.fetch(fetchRequest)
            
            // 修改数据
            handler(fetchedObjs)
            
            try context.save()
            
            if let rs = result {
                rs(nil)
            }
        } catch let error {
            if let rs = result {
                rs(error as NSError)
            }
        }
    }
    
    //MARK: - 批量操作
    
    /// 批量删除操作
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - predicate: 查找条件
    ///   - handle: 结果回调
    class func batchDelete(entity name: String, predicate: NSPredicate, resultHandler handler: ((_ error: NSError?) -> Void)? = nil) {
        
        let fetchReq = NSFetchRequest<NSFetchRequestResult>(entityName: name)
        fetchReq.predicate = predicate
        
        let batchDelete = NSBatchDeleteRequest(fetchRequest: fetchReq)
        batchDelete.resultType = .resultTypeObjectIDs
        
        do {
            let result = try context.execute(batchDelete)
            
            if let rs = result as? NSBatchDeleteResult {
                if let ids = rs.result as? Array<Any> {
                    // 告诉当前context, 数据库有变化了
                    NSManagedObjectContext.mergeChanges(fromRemoteContextSave: [NSDeletedObjectsKey: ids], into: [context])
                }
            }
            
            if let hand = handler {
                hand(nil)
            }
        } catch let error {
            
            if let hand = handler {
                hand(error as NSError)
            }
        }
    }
    
    /// 批量更新
    ///
    /// - Parameters:
    ///   - name: 实体(Entity)名称
    ///   - predicate: 查询条件
    ///   - values: 更新的值, 字典; key: 待更新属性名称, value: 更新的值
    ///   - handle: 更新结果回调
    class func batchUpdate(entity name: String, predicate: NSPredicate, propertiesToUpdate values: Dictionary<String, Any>, resultHandler handler: ((_ error: NSError?) -> Void)? = nil) {
        
        let updateReq = NSBatchUpdateRequest(entityName: name)
        updateReq.predicate = predicate
        updateReq.propertiesToUpdate = values
        updateReq.resultType = .updatedObjectIDsResultType
        
        do {
            let result = try context.execute(updateReq)
            
            if let rs = result as? NSBatchUpdateResult {
                if let ids = rs.result as? Array<NSManagedObjectID> {
                    
                    NSManagedObjectContext.mergeChanges(fromRemoteContextSave: [NSUpdatedObjectsKey: ids], into: [context])
                }
            }
            
            if let hand = handler {
                
                hand(nil)
            }
        } catch let error {
            
            if let hand = handler {
                hand(error as NSError)
            }
        }
    }
}

// MARK: - 预设的谓词查询条件 NSPredicate
enum LQPredicateType {
    
    case like(string: String, forProperty: String)
    case match(string: String, forProperty: String)
    case matchc(string: String, forProperty: String)
    case contain(string: String, forProperty: String)
    
    case equalTo(number: Int64, forProperty: String)
    case bigThan(number: Int64, forProperty: String)
    case smallThan(number: Int64, forProperty: String)
    
    case notBigThan(number: Int64, forProperty: String)
    case notSmallThan(number: Int64, forProperty: String)
}

extension CoreDataTool {
    
    class func predicate(_ type: LQPredicateType) -> NSPredicate {
        
        switch type {
        case let .like(value, key):// 将let写到前面也可以
            return NSPredicate(format: "%K LIKE %@", key, value)
        case .match(let value, let key):
            return NSPredicate(format: "%K MATCHES %@", key, value)
        case .matchc(let value, let key):
            return NSPredicate(format: "%K MATCHES[c] %@", key, value)
        case .contain(let value, let key):
            return NSPredicate(format: "%K CONTAINS %@", key, value)
        case .equalTo(let value, let key):
            return NSPredicate(format: "%K == \(value)", key)
        case .bigThan(let value, let key):
            return NSPredicate(format: "%K > \(value)", key)
        case .smallThan(let value, let key):
            return NSPredicate(format: "%K < \(value)", key)
        case .notSmallThan(let value, let key):
            return NSPredicate(format: "%K >= \(value)", key)
        case .notBigThan(let value, let key):
            return NSPredicate(format: "%K <= \(value)", key)
        }
    }
}

